### DISCUSSION TITLE
Issue with Deploying Supabase Edge Functions via Vercel

### DISCUSSION THREAD

**Discussion by Dwite** [supabase/supabase#22470] [Questions]
# Issue with Deploying Supabase Edge Functions via Vercel

## Overview

I am attempting to deploy Supabase Edge Functions via Vercel but encountering several issues that suggest a potential misalignment between Supabase's Edge Functions, which utilize Deno, and Vercel's deployment model. My main objective is to integrate Supabase's serverless functions seamlessly with my Vercel-hosted project, leveraging the benefits of edge computing. I'm seeking insights or solutions to overcome these hurdles.

## Issues Encountered

1. **TypeScript Error on Deployment:**
   - When deploying via Vercel, I receive a TypeScript error stating `Type error: Cannot find name 'Deno'`. This error suggests that the Deno runtime, which Supabase Edge Functions rely on, is not recognized in the Vercel deployment environment.

2. **Vercel Configuration for Custom Serverless Functions:**
   - Following Vercel's documentation to configure a custom environment for serverless functions does not resolve my issue. Specifically, the error message `Error: The pattern "./supabase/functions/createStory/index.ts" defined in \`functions\` doesn't match any Serverless Functions.` appears. It implies that the configuration is only applicable to Vercel's native serverless functions and does not extend support to Supabase Edge Functions structured for Deno.

## Questions

1. **Is it feasible to deploy Supabase Edge Functions via Vercel?**
   - If so, what are the correct steps or configuration changes required to ensure compatibility between Supabase's Deno runtime and Vercel's deployment model?

2. **Custom Environment Configuration:**
   - How can one configure Vercel to recognize and deploy Supabase Edge Functions properly, especially given the current limitations with the `functions` configuration in `vercel.json`?

3. **Alternatives or Workarounds:**
   - Are there any known workarounds or alternative deployment strategies that would allow for the utilization of Supabase Edge Functions within a Vercel-hosted environment?

## Additional Context

- I have followed the official documentation from both Supabase and Vercel but seem to have hit a roadblock with these specific issues.
- The primary goal is to leverage the speed and scalability of edge functions for a project that is hosted on Vercel.

I appreciate any insights, guidance, or shared experiences regarding this deployment challenge. Thank you in advance for your support!

**Comment by nischalj10** [comment]
hey @Dwite, I am stuck exactly at the same place. have you found a solution?

**Reply by nischalj10** [reply]
my current approach of solving this is adding supabase folder to exclude section in tsconfig.json
`"exclude": ["node_modules", "supabase"]`

Vercel deploys the rest of the app and the supabase edge functions have to be manually deployed. but I am sure a pre-build hook can be added to even do that automatically.

**Reply by lewiskyron** [reply]
This is what I would actually recommend if you have deployed your functions on supabase. In such cases, there is no need for you code to be bundled in the vercel build if you edge function does not  directly interact with your web application. Eg. an edge function for triggering notifications via email has no direct dependency with the web app since the edge function is mostly triggered by a webhook which is dependent on a specific table in the database.

**Comment by arnav0805** [comment]
hey @Dwite
To my best knowledge here is my answer to your question:
1.Deploying Supabase Edge Functions on Vercel is challenging because Vercel's serverless architecture uses Node.js, while Supabase's Edge Functions run on Deno. Vercel does not natively support the Deno runtime, which creates compatibility issues.
2.Since Vercel does not support Deno natively, one solution could be running Supabase Edge Functions externally and then using Vercel to route requests to them via API. Alternatively, you can self-host your edge functions or deploy them on platforms that support Deno, such as Cloudflare Workers or Supabase’s own hosting infrastructure.
3.Unfortunately, Vercel’s custom environment configuration (vercel.json) won’t resolve the Deno runtime issue. Instead, the best approach is to decouple Supabase’s Deno-based functions from Vercel and handle them separately, integrating them through API calls or utilizing middleware to manage requests between Vercel and Supabase Edge.

**Comment by mkultracreative** [comment]
> hey @Dwite
> 
> To my best knowledge here is my answer to your question:
> 
> 1.Deploying Supabase Edge Functions on Vercel is challenging because Vercel's serverless architecture uses Node.js, while Supabase's Edge Functions run on Deno. Vercel does not natively support the Deno runtime, which creates compatibility issues.
> 
> 2.Since Vercel does not support Deno natively, one solution could be running Supabase Edge Functions externally and then using Vercel to route requests to them via API. Alternatively, you can self-host your edge functions or deploy them on platforms that support Deno, such as Cloudflare Workers or Supabase’s own hosting infrastructure.
> 
> 3.Unfortunately, Vercel’s custom environment configuration (vercel.json) won’t resolve the Deno runtime issue. Instead, the best approach is to decouple Supabase’s Deno-based functions from Vercel and handle them separately, integrating them through API calls or utilizing middleware to manage requests between Vercel and Supabase Edge.

The issue is with Auth.js v5+ is that you cannot run any queries using middleware. Also it does actually work. I'm not using deno and still getting issue but the default Vercel stripe next auth template works out of the box. I think it's a matter of removing 'NEXT_PUBLIC_' from the env. It's prob an RLS policy not allowing secrets that aren't publicly exposed. I'm testing.
